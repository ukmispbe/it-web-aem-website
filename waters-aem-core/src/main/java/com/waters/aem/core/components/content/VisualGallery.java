package com.waters.aem.core.components.content;

import static com.icfolson.aem.library.core.constants.ComponentConstants.*;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.stream.Collectors;

import javax.annotation.Nonnull;
import javax.inject.Inject;

import org.apache.commons.io.FilenameUtils;
import org.apache.commons.lang3.StringUtils;
import org.apache.sling.api.SlingHttpServletRequest;
import org.apache.sling.api.resource.Resource;
import org.apache.sling.models.annotations.DefaultInjectionStrategy;
import org.apache.sling.models.annotations.Exporter;
import org.apache.sling.models.annotations.Model;
import org.apache.sling.models.annotations.injectorspecific.OSGiService;
import org.apache.sling.models.annotations.injectorspecific.ScriptVariable;
import org.apache.sling.models.annotations.injectorspecific.Self;

import com.adobe.cq.export.json.ComponentExporter;
import com.adobe.cq.export.json.ExporterConstants;
import com.citytechinc.cq.component.annotations.Component;
import com.citytechinc.cq.component.annotations.DialogField;
import com.citytechinc.cq.component.annotations.Listener;
import com.citytechinc.cq.component.annotations.Tab;
import com.citytechinc.cq.component.annotations.widgets.MultiField;
import com.citytechinc.cq.component.annotations.widgets.PathField;
import com.citytechinc.cq.component.annotations.widgets.TextField;
import com.day.cq.dam.api.Asset;
import com.day.cq.dam.api.DamConstants;
import com.day.cq.wcm.api.designer.Style;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.waters.aem.core.components.SiteContext;
import com.waters.aem.core.constants.WatersConstants;
import com.waters.aem.core.services.brightcove.BrightcoveService;
import com.waters.aem.core.servlets.ResizeImageServlet;
import com.waters.aem.core.utils.AssetUtils;
import com.waters.aem.core.utils.BrightcoveUtils;

@Component(value = "Visual Gallery",
	description = "This is the Visual Gallery component for Waters site",
    listeners = {
		@Listener(name = EVENT_AFTER_INSERT, value = REFRESH_PAGE),
		@Listener(name = EVENT_AFTER_EDIT, value = REFRESH_PAGE),
		@Listener(name = EVENT_AFTER_MOVE, value = REFRESH_PAGE),
		@Listener(name = EVENT_AFTER_COPY, value = REFRESH_PAGE),
		@Listener(name = EVENT_AFTER_DELETE, value = REFRESH_PAGE)
    },
	tabs = {
		@Tab(title = "Images"),
		@Tab(title = "Videos")
	})
@Model(adaptables = SlingHttpServletRequest.class,
    adapters = { VisualGallery.class, ComponentExporter.class },
    resourceType = VisualGallery.RESOURCE_TYPE,
    defaultInjectionStrategy = DefaultInjectionStrategy.OPTIONAL)
@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME,
    extensions = ExporterConstants.SLING_MODEL_EXTENSION)

public final class VisualGallery implements ComponentExporter {

	public static final String RESOURCE_TYPE = "waters/components/content/visualgallery";
	
	public String[] tabs = {"Images", "Videos"};
	
	 private static final ObjectMapper MAPPER = new ObjectMapper();
	
	 private static final String SRC_URI_TEMPLATE_WIDTH = "{{width}}";

	@OSGiService
	private BrightcoveService brightcoveService;

	@Self
	private SiteContext siteContext;

	@Inject
	private Resource resource;

	@ScriptVariable
	private Style currentStyle;


	@DialogField(fieldLabel = "Images",
			fieldDescription = "Select Images to be displayed",
			renderReadOnly = false,
			tab = 1,
			ranking = 1)
	@MultiField
	@PathField(rootPath = WatersConstants.DAM_PATH)
	@Inject
	private String[] images = new String[0];


	@DialogField(fieldLabel = "Video ID",
			fieldDescription = "Enter Brightcove Video ID that needs to be displayed in the video gallery",
			tab = 2,
			ranking = 1)
	@MultiField
	@TextField
	@Inject
	private String[] videoIds = new String[0];


	public String getImageGalleryAsJson() throws JsonProcessingException {
		final Map<String, Object> jsonMap = new HashMap<>();
		jsonMap.put("images", getAssetsList());
		jsonMap.put("widths", getWidths());
		jsonMap.put("videoIds", getVideoIds());
		jsonMap.put("brightcoveAccount", getBrightcoveAccount());
		jsonMap.put("brightcovePlayerId", getBrightcovePlayerId());
		jsonMap.put("tabs", getTabs());

		return MAPPER.writeValueAsString(jsonMap);
	}

	private List<Map<String, String>> getAssetsList() {

		List<Map<String, String>> assetsList = null;

		if (images.length != 0) {
			assetsList = new ArrayList<>();
			for (String image : images) {
				Map<String, String> imageMap = new HashMap<>();
				Asset asset = AssetUtils.getAsset(resource.getResourceResolver(), image);
				String alt = AssetUtils.getAltText(asset);
				//In Build URI method also we are doing this in case of false
				imageMap.put("src", buildUri(asset, true));		
				imageMap.put("title", asset.getMetadataValue(DamConstants.DC_TITLE) == null ? StringUtils.EMPTY
						: asset.getMetadataValue(DamConstants.DC_TITLE));
				imageMap.put("alt", alt == null ? StringUtils.EMPTY : alt);	
				imageMap.put("description",
						asset.getMetadataValue(DamConstants.DC_DESCRIPTION) == null ? StringUtils.EMPTY
								: asset.getMetadataValue(DamConstants.DC_DESCRIPTION));

				assetsList.add(imageMap);
			}

			return assetsList.stream().filter(Objects::nonNull).collect(Collectors.toList());
		}else {
			return assetsList;
		}
	}

	private List<String> getWidths() {
		return Arrays.asList(currentStyle.get(Image.PN_DESIGN_ALLOWED_RENDITION_WIDTHS, new String[0]));
	}
	
    private String buildUri(final Asset asset, final boolean template) {
        final StringBuilder builder = new StringBuilder();

        // append the DAM asset path
        builder.append(asset.getPath());

        if (template) {
            builder.append(".");

            // width selector
            builder.append(SRC_URI_TEMPLATE_WIDTH);
            builder.append(".");

            // 'resize' extension to resolve the resize image servlet
            builder.append(ResizeImageServlet.RESIZE_EXTENSION);
            builder.append("/");
            builder.append(ResizeImageServlet.SUFFIX_NAME);
            builder.append(".");

            // add the extension derived from DAM asset name
            builder.append(FilenameUtils.getExtension(asset.getName()));
        }

        return builder.toString();
    }

	public String[] getVideoIds() {
		return videoIds;
	}
	
	public String[] getTabs() {
		return tabs;
	}

	public String getBrightcoveAccount() {
		return BrightcoveUtils.getBrightcoveAccount(siteContext, brightcoveService);
	}

	public String getBrightcovePlayerId() {
		return BrightcoveUtils.getBrightcovePlayerId(siteContext, brightcoveService);
	}
	
	@Nonnull
	@Override
	public String getExportedType() {
		return RESOURCE_TYPE;
	}
}