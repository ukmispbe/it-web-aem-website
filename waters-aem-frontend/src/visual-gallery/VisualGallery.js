import React, { useState, useEffect } from "react";
import PropTypes from "prop-types";
import { getBrightCoveVideoData } from "./services";
import OverlayWithTabs from "../overlay-with-tabs/OverlayWithTabs";
import ImageGallery from "./overlay/OverlayImageGallery";
import VideoGallery from "./overlay/OverlayVideoGallery";
import MobileOverlay from "./overlay/MobileOverlay";
import { WIDTHS } from "../constants";
import { msToMinAndSeconds } from "../utils/userFunctions";

const VisualGallery = ({
  tabs,
  templates,
  videoIds,
  brightcoveAccount,
  brightcovePlayerId,
  widths,
  zoomLabel,
  zoomInIcon,
  brightCoveApi,
  policyKey,
}) => {
  // dummy data for mobile overlay.. will be deleted later on
  const overlayProps = {
    iconClear:
      "https://dev1.waters.com/content/dam/waters/en/brand-assets/icons/close.svg",
    iconLeft:
      "https://dev1.waters.com/content/dam/waters/en/brand-assets/icons/left.svg",
    backBtnLabel: "Back",
    closeBtnLabel: "Close",
  };
  // dummy video data for mobile overlay.. will be removed later once API integration done
  const [videoConfig, setVideoConfig] = useState([
    {
      brightcoveVideoId: videoIds[0],
      brightcoveAccount,
      brightcovePlayerId,
      duration: msToMinAndSeconds(789654),
      title: "Video Title",
      description: "Video Description",
      brightcovePlayerId,
      isVideo: true,
      showVideoPlayButton: true,
      videoIconURL:
        "https://dev1.waters.com/content/dam/waters/en/brand-assets/icons/play.svg",
      thumbnail:
        "http://localhost:4502/content/dam/waters/emails/innovations-acquity-premier-column-logo.jpeg",
    },
    {
      brightcoveVideoId: videoIds[1],
      brightcoveAccount,
      brightcovePlayerId,
      duration: msToMinAndSeconds(389654),
      title: "Video Title two",
      description: "Video Description two",
      brightcovePlayerId,
      videoIconURL:
        "https://dev1.waters.com/content/dam/waters/en/brand-assets/icons/play.svg",
      thumbnail:
        "http://localhost:4502/content/dam/waters/emails/innovations-otto-spe-logo.jpeg",
    },
  ]);
  const [activeTabIndex, setActiveTabIndex] = useState(0);
  const [isOpen, setIsOpen] = useState(true);
  const [brightCoveData, setBrightCoveData] = useState([]);
  const [isMobileOverlayOpen, mobileOverlayHandler] = useState(false);

  const handleClose = (e) => setIsOpen(false);

  const activeTabHandler = (id) => setActiveTabIndex(id);
  /* will be deleted later on */
  const launchMobileOverlay = () => {
    mobileOverlayHandler(!isMobileOverlayOpen);
  };

  const brightCoveDataConstruct = (data) => {
    const videoObj = data.map((obj) => ({
      title: obj.name,
      description: obj.description,
      thumbnail: obj.thumbnail,
      duration: obj.duration,
      brightcoveVideoId: obj.id,
      brightcoveAccount,
      brightcovePlayerId,
    }));
    setBrightCoveData(videoObj);
  };

  useEffect(() => {
    getBrightCoveVideoData(
      videoIds,
      brightcoveAccount,
      brightCoveApi,
      policyKey,
      brightCoveDataConstruct
    );
  }, []);

  const closeMobileOverlay = () => {
    mobileOverlayHandler(false);
  };
  return (
    <div>
      {/* This button will remove later it is just to launch mobile overlay */}
      <button onClick={launchMobileOverlay}>{`Launch Mobile Overlay`}</button>
      <OverlayWithTabs
        isOpen={isOpen}
        tabs={tabs}
        handleClose={handleClose}
        activeTabHandler={activeTabHandler}
      >
        <div>
          {templates.length > 0 && (
            <div style={{ display: activeTabIndex === 0 ? "block" : "none" }}>
              <ImageGallery
                templates={templates}
                widths={widths}
                zoomInIcon={zoomInIcon}
                zoomLabel={zoomLabel}
              />
            </div>
          )}
          {brightCoveData.length > 0 && (
            <div style={{ display: activeTabIndex === 1 ? "block" : "none" }}>
              <VideoGallery brightCoveData={brightCoveData} />
            </div>
          )}
        </div>
      </OverlayWithTabs>
      {isMobileOverlayOpen && (
        <MobileOverlay
          iconClear={overlayProps.iconClear}
          iconLeft={overlayProps.iconLeft}
          backBtnLabel={overlayProps.backBtnLabel}
          closeBtnLabel={overlayProps.closeBtnLabel}
          closeMobileOverlay={closeMobileOverlay}
          templates={templates}
          widths={widths}
          zoomInIcon={zoomInIcon}
          zoomLabel={zoomLabel}
          videoConfig={videoConfig}
        ></MobileOverlay>
      )}
    </div>
  );
};

VisualGallery.defaultProps = {
  widths: WIDTHS,
  alt: "",
  templates: [],
  videoConfig: [],
};

VisualGallery.propTypes = {
  alt: PropTypes.string,
  templates: PropTypes.arrayOf(PropTypes.string).isRequired,
  videoConfig: PropTypes.arrayOf(PropTypes.object),
  widths: PropTypes.arrayOf(PropTypes.string).isRequired,
  zoomInIcon: PropTypes.string.isRequired,
  tabs: PropTypes.arrayOf(PropTypes.string),
};

export default VisualGallery;
