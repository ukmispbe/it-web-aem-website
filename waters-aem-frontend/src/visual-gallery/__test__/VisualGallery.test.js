import React from "react";
import { shallow } from "enzyme";
import VisualGallery from "../VisualGallery";
const defaultConfig = {
  templates: [
    {
      src:
        "http://localhost:4502/content/dam/waters/en/brand-assets/product/covid19/banner-covid-innovation.jpg.{{width}}.resize/img.jpg",
      alt: "test",
      description: "test",
      title: "test",
    },
    {
      src:
        "http://localhost:4502/content/dam/waters/en/brand-assets/product/catalogs/catalog-2019-2020.jpg.{{width}}.resize/img.jpg",
      alt: "alt text",
      description: "Some description",
      title: "test",
    },
  ],
  widths: ["128", "450", "650", "789", "1280"],
  zoomInIcon:
    "http://localhost:4502/content/dam/waters/en/brand-assets/icons/zoom-in.svg",

  alt: "test",
  videoIds: ["1234566", "34343434"],
  tabs: ["Images", "Videos"],
  zoomLabel: "Tap or Click to Zoom",
  brightcoveAccount: "4324324324324",
  brightcovePlayerId: "3434234324324",
};

describe("<VisualGallery />", () => {
  let VisualGalleryComponent = "";
  beforeEach(() => {
    VisualGalleryComponent = shallow(<VisualGallery {...defaultConfig} />);
  });

  test("should render correctly", () => {
    expect(VisualGalleryComponent).toMatchSnapshot();
  });

  test("should render Image Gallery with no tabs", () => {
    const props = {
      templates: [
        {
          src:
            "http://localhost:4502/content/dam/waters/en/brand-assets/product/covid19/banner-covid-innovation.jpg.{{width}}.resize/img.jpg",
          alt: "test",
          description: "test",
          title: "test",
        },
      ],
      widths: ["128", "450", "650", "789", "1280"],
      zoomInIcon:
        "http://localhost:4502/content/dam/waters/en/brand-assets/icons/zoom-in.svg",

      alt: "test",
      videoIds: ["1234566", "34343434"],
      tabs: ["Images"],
      zoomLabel: "Tap or Click to Zoom",
      brightcoveAccount: "4324324324324",
      brightcovePlayerId: "3434234324324",
    };
    const wrapper = shallow(<VisualGallery {...props} />);
    expect(wrapper).toMatchSnapshot();
  });
});
