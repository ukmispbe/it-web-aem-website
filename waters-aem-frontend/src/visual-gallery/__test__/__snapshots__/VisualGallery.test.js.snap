// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VisualGallery /> should render Image Gallery with no tabs 1`] = `ShallowWrapper {}`;

exports[`<VisualGallery /> should render correctly 1`] = `ShallowWrapper {}`;
